File: Message.js
const mongoose = require('mongoose');

const messageSchema = new mongoose.Schema({
  phoneNumber: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'PhoneNumber',
    required: [true, 'Phone number is required']
  },
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'User is required']
  },
  fromNumber: {
    type: String,
    required: [true, 'Sender number is required'],
    trim: true
  },
  toNumber: {
    type: String,
    required: [true, 'Recipient number is required'],
    trim: true
  },
  content: {
    type: String,
    required: [true, 'Message content is required'],
    maxlength: [1600, 'Message content cannot exceed 1600 characters']
  },
  messageType: {
    type: String,
    enum: ['sms', 'mms', 'otp', 'verification', 'notification', 'other'],
    default: 'sms'
  },
  detectedOTP: {
    type: String,
    default: null
  },
  provider: {
    type: String,
    required: [true, 'Provider is required'],
    enum: ['twilio', 'nexmo', 'vonage', 'other']
  },
  providerMessageId: {
    type: String,
    required: false
  },
  isRead: {
    type: Boolean,
    default: false
  },
  readAt: {
    type: Date,
    default: null
  },
  receivedAt: {
    type: Date,
    default: Date.now
  },
  metadata: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  attachments: [{
    type: {
      type: String,
      enum: ['image', 'video', 'audio', 'document']
    },
    url: String,
    filename: String,
    size: Number,
    mimeType: String
  }]
}, {
  timestamps: true,
  toJSON: {
    transform: function(doc, ret) {
      delete ret.__v;
      return ret;
    }
  }
});

// Indexes for better query performance
messageSchema.index({ phoneNumber: 1, receivedAt: -1 });
messageSchema.index({ user: 1, receivedAt: -1 });
messageSchema.index({ isRead: 1 });
messageSchema.index({ messageType: 1 });
messageSchema.index({ receivedAt: -1 });
messageSchema.index({ fromNumber: 1 });

// Virtual for message age
messageSchema.virtual('age').get(function() {
  return Date.now() - this.receivedAt.getTime();
});

// Virtual for formatted received time
messageSchema.virtual('formattedReceivedAt').get(function() {
  return this.receivedAt.toLocaleString();
});

// Instance method to mark as read
messageSchema.methods.markAsRead = function() {
  this.isRead = true;
  this.readAt = new Date();
  return this.save();
};

// Instance method to detect OTP
messageSchema.methods.detectOTP = function() {
  const content = this.content.toLowerCase();
  
  // Common OTP patterns
  const otpPatterns = [
    /\b(\d{4,8})\b/g,  // 4-8 digit codes
    /code[:\s]*(\d+)/gi,  // "code: 123456"
    /verification[:\s]*(\d+)/gi,  // "verification: 123456"
    /otp[:\s]*(\d+)/gi,  // "otp: 123456"
    /pin[:\s]*(\d+)/gi,  // "pin: 123456"
    /token[:\s]*(\d+)/gi,  // "token: 123456"
  ];
  
  // Keywords that suggest this is a verification message
  const verificationKeywords = [
    'verification', 'verify', 'confirm', 'authenticate', 
    'code', 'otp', 'pin', 'security', 'login', 'signin',
    'access', 'account', 'password', 'reset'
  ];
  
  // Check if content contains verification keywords
  const hasVerificationKeywords = verificationKeywords.some(keyword => 
    content.includes(keyword)
  );
  
  if (hasVerificationKeywords) {
    // Try to extract OTP code
    for (const pattern of otpPatterns) {
      const matches = this.content.match(pattern);
      if (matches) {
        // Get the first numeric match
        const numericMatch = matches.find(match => /^\d+$/.test(match.replace(/\D/g, '')));
        if (numericMatch) {
          const code = numericMatch.replace(/\D/g, '');
          if (code.length >= 4 && code.length <= 8) {
            this.detectedOTP = code;
            this.messageType = 'otp';
            return code;
          }
        }
      }
    }
    
    // If we found verification keywords but no clear OTP, mark as verification
    this.messageType = 'verification';
  }
  
  return null;
};

// Instance method to get formatted content with OTP highlighting
messageSchema.methods.getFormattedContent = function() {
  let content = this.content;
  
  if (this.detectedOTP) {
    // Highlight the OTP code in the content
    const otpRegex = new RegExp(`\\b${this.detectedOTP}\\b`, 'g');
    content = content.replace(otpRegex, `**${this.detectedOTP}**`);
  }
  
  return content;
};

// Static method to create message from webhook
messageSchema.statics.createFromWebhook = async function(data) {
  const PhoneNumber = mongoose.model('PhoneNumber');
  
  // Find the phone number
  const phoneNumber = await PhoneNumber.findOne({ 
    phoneNumber: data.toNumber 
  }).populate('user');
  
  if (!phoneNumber || !phoneNumber.user) {
    throw new Error('Phone number not found or not assigned to user');
  }
  
  // Create message
  const message = new this({
    phoneNumber: phoneNumber._id,
    user: phoneNumber.user._id,
    fromNumber: data.fromNumber,
    toNumber: data.toNumber,
    content: data.content,
    provider: data.provider,
    providerMessageId: data.providerMessageId,
    metadata: data.metadata || {}
  });
  
  // Detect OTP
  message.detectOTP();
  
  await message.save();
  
  // Add message to phone number's messages array
  phoneNumber.messages.push(message._id);
  await phoneNumber.save();
  
  return message.populate(['phoneNumber', 'user']);
};

// Static method to find messages for user
messageSchema.statics.findByUser = function(userId, options = {}) {
  const {
    phoneNumberId,
    messageType,
    isRead,
    since,
    limit = 50,
    offset = 0,
    sort = { receivedAt: -1 }
  } = options;
  
  const query = { user: userId };
  
  if (phoneNumberId) query.phoneNumber = phoneNumberId;
  if (messageType) query.messageType = messageType;
  if (isRead !== undefined) query.isRead = isRead;
  if (since) query.receivedAt = { $gte: since };
  
  return this.find(query)
    .populate('phoneNumber', 'phoneNumber formattedNumber')
    .sort(sort)
    .skip(offset)
    .limit(limit);
};

// Static method to mark all messages as read for a user
messageSchema.statics.markAllAsReadForUser = async function(userId, phoneNumberId = null) {
  const query = { user: userId, isRead: false };
  if (phoneNumberId) query.phoneNumber = phoneNumberId;
  
  const result = await this.updateMany(query, {
    isRead: true,
    readAt: new Date()
  });
  
  return result.modifiedCount;
};

// Static method to cleanup old messages
messageSchema.statics.cleanupOldMessages = async function(daysOld = 30) {
  const cutoffDate = new Date(Date.now() - (daysOld * 24 * 60 * 60 * 1000));
  
  const result = await this.deleteMany({
    receivedAt: { $lt: cutoffDate }
  });
  
  return result.deletedCount;
};

// Pre-save middleware to auto-detect message type
messageSchema.pre('save', function(next) {
  if (this.isNew && !this.detectedOTP) {
    this.detectOTP();
  }
  next();
});

module.exports = mongoose.model('Message', messageSchema);

File: message.py
from datetime import datetime
from datetime import timedelta
import re
from src.models.user import db

class Message(db.Model):
    __tablename__ = 'messages'
    
    id = db.Column(db.Integer, primary_key=True)
    phone_number_id = db.Column(db.Integer, db.ForeignKey('phone_numbers.id'), nullable=False)
    sender_number = db.Column(db.String(20), nullable=False)
    message_content = db.Column(db.Text, nullable=False)
    message_type = db.Column(db.String(20), default='sms', nullable=False)
    received_at = db.Column(db.DateTime, default=datetime.utcnow)
    provider_message_id = db.Column(db.String(255), nullable=True)
    provider_id = db.Column(db.Integer, db.ForeignKey('sms_providers.id'), nullable=True)
    is_read = db.Column(db.Boolean, default=False)
    extra_data = db.Column(db.JSON, default={})
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    
    # Relationships
    provider = db.relationship('SMSProvider', backref='messages', lazy=True)
    
    def __repr__(self):
        return f'<Message {self.id} from {self.sender_number}>'
    
    def to_dict(self):
        return {
            'id': self.id,
            'phone_number_id': self.phone_number_id,
            'phone_number': self.phone_number.phone_number if self.phone_number else None,
            'sender_number': self.sender_number,
            'message_content': self.message_content,
            'message_type': self.message_type,
            'received_at': self.received_at.isoformat(),
            'is_read': self.is_read,
            'metadata': self.extra_data
        }
    
    def mark_as_read(self):
        """Mark this message as read"""
        self.is_read = True
        db.session.commit()
    
    def detect_message_type(self):
        """Automatically detect message type based on content"""
        content = self.message_content.lower()
        
        # OTP detection patterns
        otp_patterns = [
            r'\b\d{4,8}\b',  # 4-8 digit codes
            r'code[:\s]*\d+',  # "code: 123456"
            r'verification[:\s]*\d+',  # "verification: 123456"
            r'otp[:\s]*\d+',  # "otp: 123456"
        ]
        
        verification_keywords = [
            'verification', 'verify', 'confirm', 'authenticate', 
            'code', 'otp', 'pin', 'security'
        ]
        
        # Check for OTP patterns
        for pattern in otp_patterns:
            if re.search(pattern, content):
                if any(keyword in content for keyword in verification_keywords):
                    return 'otp'
        
        # Check for verification keywords
        if any(keyword in content for keyword in verification_keywords):
            return 'verification'
        
        return 'sms'
    
    @classmethod
    def create_from_webhook(cls, phone_number_id, sender_number, content, provider_id=None, provider_message_id=None):
        """Create a new message from webhook data"""
        message = cls(
            phone_number_id=phone_number_id,
            sender_number=sender_number,
            message_content=content,
            provider_id=provider_id,
            provider_message_id=provider_message_id
        )
        
        # Auto-detect message type
        message.message_type = message.detect_message_type()
        
        db.session.add(message)
        db.session.commit()
        return message
    
    @classmethod
    def get_messages_for_user(cls, user_id, phone_number_id=None, message_type=None, is_read=None, since=None, limit=50, offset=0):
        """Get messages for a specific user with filtering options"""
        from src.models.phone_number import PhoneNumber
        
        # Join with phone_numbers to filter by user
        query = cls.query.join(PhoneNumber).filter(PhoneNumber.user_id == user_id)
        
        if phone_number_id:
            query = query.filter(cls.phone_number_id == phone_number_id)
        
        if message_type:
            query = query.filter(cls.message_type == message_type)
        
        if is_read is not None:
            query = query.filter(cls.is_read == is_read)
        
        if since:
            query = query.filter(cls.received_at >= since)
        
        return query.order_by(cls.received_at.desc()).offset(offset).limit(limit).all()
    
    @classmethod
    def cleanup_old_messages(cls, hours=24):
        """Clean up messages older than specified hours"""
        cutoff_time = datetime.utcnow() - timedelta(hours=hours)
        old_messages = cls.query.filter(cls.received_at < cutoff_time).all()
        
        for message in old_messages:
            db.session.delete(message)
        
        db.session.commit()
        return len(old_messages)

